[DOC] Tutorial for semantic segmentation (#1766)
open-edge-platform/training_extensions: docs/source/guide/tutorials/base/how_to_train/semantic_segmentation.rst

  |     TASK     |                          ID                         |        NAME        |                                BASE PATH                                 |
  +--------------+-----------------------------------------------------+--------------------+--------------------------------------------------------------------------+
  | SEGMENTATION |    Custom_Semantic_Segmentation_Lite-HRNet-18_OCR   |   Lite-HRNet-18    |   otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/template.yaml    |
  | SEGMENTATION | Custom_Semantic_Segmentation_Lite-HRNet-18-mod2_OCR | Lite-HRNet-18-mod2 | otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml |
  | SEGMENTATION |  Custom_Semantic_Segmentation_Lite-HRNet-s-mod2_OCR | Lite-HRNet-s-mod2  | otx/algorithms/segmentation/configs/ocr_lite_hrnet_s_mod2/template.yaml  |
  | SEGMENTATION |  Custom_Semantic_Segmentation_Lite-HRNet-x-mod3_OCR | Lite-HRNet-x-mod3  | otx/algorithms/segmentation/configs/ocr_lite_hrnet_x_mod3/template.yaml  |
  +--------------+-----------------------------------------------------+--------------------+--------------------------------------------------------------------------+

> Note that we do not attach OCR head for supported models in default. We remain the suffix '_OCR' in ID just for backward compatibility.

Reviewer: Emphasizing this sentence is more beneficial to users.

"Note that we do not attach OCR head for supported models in default."

Author: I made it NOTE!!

> Note that we do not attach OCR head for supported models in default. We remain the suffix '_OCR' in ID just for backward compatibility.

Reviewer 2: ```suggestion
.. note::

    We do not attach OCR head for supported models in default. We remain the suffix '_OCR' in ID just for backward compatibility.
```

To have a specific example in this tutorial, all commands will be run on the :ref:`Lite-HRNet-18-mod2 <semantic_segmentation_models>`  model. It's a light model, that achieves competitive accuracy while keeping the inference fast.

2.  Next, we need to create train/validation sets. OTX supports auto-split functionality for the semantic segmentation.

Let's prepare an OTX semantic segmentation workspase running the following command:

.. note::

>   Currently, OTX supports auto-split only for public VOC dataset format in semantic segmentation. We should specify the validation roots in argument '--val-data-roots' when using custom segmentation dataset. About custom segmentation dataset, please refer to the :doc:`advanced tutorial for datumaro <../../advanced/datumaro>`.

Reviewer 2: ```suggestion
  Currently, OTX supports auto-split only for public VOC dataset format in semantic segmentation. We should specify the validation roots in argument '--val-data-roots' when using other supported segmentation dataset. About dataset formats for semantic segmentation, please refer to the :doc:`explanation section <../../../explanation/algorithms/segmentation/semantic_segmentation>`.
```

  [*] Load Model Name: Lite-HRNet-18-mod2

  ...

  [*] Update data configuration file to: otx-workspace-SEGMENTATION/data.yaml

  (otx) ...$ cd ./otx-workspace-SEGMENTATION

It will create **otx-workspace-SEGMENTATION** with all necessery configs for Lite-HRNet-18-mod2, prepared ``data.yaml`` to simplify CLI commands launch and splitted dataset.

2. To start training we need to call ``otx train``
command in our worspace:

.. code-block::

  (otx) ...$ otx train

That's it! The training will return artifacts: ``weights.pth`` and ``label_schema.json``, which are needed as input for the further commands: ``export``, ``eval``,  ``optimize``,  etc.

After that, we have the PyTorch model trained with OTX, which we can use for evaluation, export, optimization and deployment.

***********
Validation
***********

1. ``otx eval`` runs evaluation of a trained
model on a specific dataset.

The eval function receives test annotation information and model snapshot, trained in the previous step.
Please note, ``label_schema.json`` file contains meta-information about the dataset and it should be located in the same folder as the model snapshot.

> ``otx eval`` will output a mDice score for semantic segmentation.

Reviewer 2: ```suggestion
``otx eval`` will output a ``mDice`` score for semantic segmentation.
```

  2023-02-21 18:09:57,807 | INFO : called evaluate()
  2023-02-21 18:09:57,807 | INFO : Computing mDice
  2023-02-21 18:09:58,508 | INFO : mDice after evaluation: 0.9659400544959128
  Performance(score: 0.9659400544959128, dashboard: (1 metric groups))

In ``performance.json`` file, the validation output score is saved as:

.. code-block::

  {"Dice Average": 0.9659400544959128}

*********
Export
*********

1. ``otx export`` exports a trained Pytorch `.pth` model to the OpenVINO™ Intermediate Representation (IR) format.
It allows running the model on the Intel hardware much more efficient, especially on the CPU. Also, the resulting IR model is required to run POT optimization. IR model consists of 2 files: ``openvino.xml`` for weights and ``openvino.bin`` for architecture.

2. We can run the below command line to export the trained model
and save the exported model to the ``openvino_model`` folder.

.. code-block::

  (otx) ...$ otx export --load-weights models/weights.pth \
                        --save-model-to openvino_model

  ...

  2023-02-02 03:23:03,057 | INFO : run task done.
  2023-02-02 03:23:03,064 | INFO : Exporting completed

> 3. We can check the mDice score of the IR model and the consistency between the exported model and the PyTorch model,

Reviewer 2: ```suggestion
3. We can check the ``mDice`` score of the IR model and the consistency between the exported model and the PyTorch model,
```

.. code-block::

  (otx) ...$ otx eval --test-data-roots splitted_dataset/val \
                      --load-weights openvino_model/openvino.xml \
                      --save-performance openvino_model/performance.json

  ...

  Performance(score: 0.9659400544959128, dashboard: (1 metric groups))

*************
Optimization
*************

1. We can further optimize the model with ``otx optimize``.
It uses NNCF or POT depending on the model format.

Please, refer to :doc:`optimization explanation <../../../explanation/additional_features/models_optimization>` section to get the intuition of what we use under the hood for optimization purposes.

2. Command example for optimizing
a PyTorch model (`.pth`) with OpenVINO™ NNCF.

.. code-block::

  (otx) ...$ otx optimize --load-weights models/weights.pth --save-model-to nncf_model

  ...

  INFO:nncf:Loaded 5286/5286 parameters
  2023-02-21 18:09:56,134 | INFO : run task done.
  2023-02-21 18:09:57,807 | INFO : called evaluate()
  2023-02-21 18:09:57,807 | INFO : Computing mDice
  2023-02-21 18:09:58,508 | INFO : mDice after evaluation: 0.9659400544959128
  Performance(score: 0.9659400544959128, dashboard: (1 metric groups))

> The optimization time relies on the hardware characteristics, for example on 1 GeForce 3090 and Intel(R) Core(TM) i9-10980XE it took about 15 minutes.

Reviewer 3: ```suggestion
The optimization time relies on the hardware characteristics, for example on 1 NVIDIA GeForce RTX 3090 and Intel(R) Core(TM) i9-10980XE it took about 15 minutes.
```

Author: I would change but seems like align it with classification and detection.

Reviewer 3: Classification and detection are wrong. Official name is `NVIDIA GeForce RTX 3090`.

Author: I didn't say this is wrong. I meant cls and det should be changed too..
